Fix `confirmed` being ignored in htmx:confirm event

matiboy opened the pull request:
Referring to #1588 

- Adds test for existing `hx-confirm` behavior by stubbing window.confirm; tests were added to a new test file `hx-confirm.js`
- Enhance `htmx:confirm` event detail to include the value found in `hx-confirm` (set to `detail.question`)
- Allow passing `skipConfirmation` to `detail.issueRequest`; when true, this skips window.confirm and triggers the request right away
- Fix the missing check for `confirmed` in original code (that was the original issue)
- Fix typo in [original `htmx:config` test ](https://github.com/bigskysoftware/htmx/blob/2534a970b6d606cb590edd8f6097e4a6440d8d06/test/core/events.js#L646) (incorrect handler would be removed in case of test failure)

Backwards compatibility: if `skipConfirmation` is not passed, the behaviour remains the same as it was previously, meaning it still calls window.confirm if `hx-confirm` has a value; I think it's incorrect, but it is the old behavior, so this should be backwards compatible

Closes #1588

alexpetros reviewed (commented):
Cleverly done. I loove the test-to-code ratio and appreciate the attention to backwards compat. Could you update the docs to indicate the presence of `detail.question` and the `skipConfirmation` argument?

I'm actually a little torn on how the `skipConfirmation` should work (I'm not sure we need that be backwards compatible, but it's always better to be on the safe side).

matiboy commented:
@alexpetros I changed the example slightly because the example given as text didn't match the one that was demo-ed at the bottom. So I kept the bottom version (uses click+custom event) as hyperscript and moved the version using `invokeRequest` to vanilla Javascript. Hope that's ok

faizanakram99 commented:
Not directly related to this pr, but why doesn't `hx-on::confirm` on a parent element work (read would it work after this PR)?

I'd like to avoid any custom JabbaScript and still use a different confirmation than window.confirm.

alexpetros commented:
> Not directly related to this pr, but why doesn't hx-on::confirm on a parent element work (read would it work after this PR)?

`hx-on` is not inherited, if that's what you mean

faizanakram99 commented:
> > Not directly related to this pr, but why doesn't hx-on::confirm on a parent element work (read would it work after this PR)?
> 
> `hx-on` is not inherited, if that's what you mean

it doesn't even work on the element that triggers event, in short how to use hx-on::confirm with hx-get/post etc to use sweetalert for example?

matiboy commented:
> > Not directly related to this pr, but why doesn't hx-on::confirm on a parent element work (read would it work after this PR)?
> 
> `hx-on` is not inherited, if that's what you mean

Note that `hx-confirm` on a parent element does work since we use `getClosestAttributeValue` in the code

matiboy commented:
> > > Not directly related to this pr, but why doesn't hx-on::confirm on a parent element work (read would it work after this PR)?
> > 
> > 
> > `hx-on` is not inherited, if that's what you mean
> 
> it doesn't even work on the element that triggers event, in short how to use hx-on::confirm with hx-get/post etc to use sweetalert for example?

Hi @faizanakram99 I've created a few examples [in this Codepen](https://codepen.io/matiboy/pen/oNQOdRp) it shows that you can use `hx-on=` or `hx-on::confirm` successfully. 

It also shows an example of how to do Sweetalert without hyperscript. However this is before my PR so you have to use the same trick as the [example available on the docs](https://htmx.org/examples/confirm/) (look at the code on the demo button, not the code in text) which is to react on click, then trigger the `hx-trigger` event of the button.

1cg commented:
Thank you @matiboy can I ask you to add some docs on the hx-confirm.md page as well?

matiboy commented:
> Thank you @matiboy can I ask you to add some docs on the hx-confirm.md page as well?

Hi @1cg the PR already contains the edits to `hx-confirm.md` and a revamp of the `examples/confirm.md` to reflect the changes.

Please let me know if you need me to add anything else.

hakosl commented:
The vanilla js example at: [htmx.org/examples/confirm](https://htmx.org/examples/confirm/) seems to trigger both the native alert(), as well as the Swal alert, and i seems to also be having the same issue locally.


<img width="718" alt="image" src="https://github.com/bigskysoftware/htmx/assets/10966424/4ed368e0-8b67-45f3-bdf4-1a54b6b4af56">

Additionally(I don't know if it's related to this PR) I am having issues with the "htmx:confirm" event being triggered on other regular events, like clicking a hx-boost link

1cg commented:
Dang it. I will roll this change back and cut a new release tonight. Sorry about that.

Telroshan commented:
Ok so it seems we just have a problem about our release script, as the website's htmx file seems to be in version 1.9.5, i.e. before that PR was merged (thus before the fix), as it came in the latest release (1.9.7).
If you run it locally, you should get the expected behavior
Investigating on the release process, but the good news is we shouldn't need to revert those changes 😄

matiboy commented:
As the author of the PR, please let me know if there is indeed an issue with the change and whether you need me to investigate further please; thanks

matiboy commented:
> Additionally(I don't know if it's related to this PR) I am having issues with the "htmx:confirm" event being triggered on other regular events, like clicking a hx-boost link

Hi @hakosl could you give a quick codepen repro or describe a bit more so I can investigate this issue please? Thanks!

matiboy commented:
Oh I think I see what you're referring to @hakosl Yes `htmx:confirm` actually triggers on everything (somewhat surprisingly) but that behavior existed before the PR. When I fixed this bug and proposed to enhance the `details` prop, I did not suggest to change that behavior since it would be breaking.

1cg commented:
no worries it was a bug in our web gen script that got the version of htmx on the website out of sync with the /arc directory

Nothing to do with your change sorry for the alarm!

Sent from [Proton Mail](https://proton.me/mail/home) for iOS

On Mon, Nov 6, 2023 at 5:15 PM, matiboy ***@***.***(mailto:On Mon, Nov 6, 2023 at 5:15 PM, matiboy <<a href=)> wrote:

> Oh I think I see what you're referring to ***@***.***(https://github.com/hakosl) Yes htmx:confirm actually triggers on everything (somewhat surprisingly) but that behavior existed before the PR. When I fixed this bug and proposed to enhance the details prop, I did not suggest to change that behavior since it would be breaking.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bigskysoftware/htmx/pull/1610#issuecomment-1797067642), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AADSRP6HG4MJFO6BQA36QCLYDF4RBAVCNFSM6AAAAAA2PABC22VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJXGA3DONRUGI).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

hakosl commented:
Agreed, with htmx 1.9.7, it works as expected the. "htmx:confirmed" event is still being triggered on all elements, but thats a different issue. 
[https://codepen.io/h-kon-osland/pen/JjxEawK](https://codepen.io/h-kon-osland/pen/JjxEawK)

abe-101 commented:
> Oh I think I see what you're referring to @hakosl Yes `htmx:confirm` actually triggers on everything (somewhat surprisingly) but that behavior existed before the PR.

Any workaround for this bug?

abe-101 commented:
I came up with this solution:

```js
document.addEventListener("htmx:confirm", function(e) {
    // Check if the question is null or empty
    if (!e.detail.question) {
        // If it is, issue the request directly
        e.detail.issueRequest(true); // use true to skip window.confirm
    } else {
        // If not, show the Swal confirmation dialog
        e.preventDefault();
        Swal.fire({
            title: "Proceed?",
            text: `${e.detail.question}`,
            icon: "warning",
            showCancelButton: true,
            confirmButtonText: "Yes",
            cancelButtonText: "No",
            reverseButtons: true
        }).then(function(result) {
            if(result.isConfirmed) e.detail.issueRequest(true); // use true to skip window.confirm
        });
    }
});
```

c-nv-s commented:
so is there a ticket anywhere for the fact ``htmx:confirm`` is triggering on everything? which is buggy behavior, especially with the documented sweetalert example if you have other htmx elements on the page that are triggered ``hx-get="/something" hx-trigger="load"`` then it triggers the sweet alert confirm message on every page load.

1cg commented:
the `htmx:confirm` event is always triggered on every request.  this is by design: it provides a scripting mechanism to cancel or delay a request that is integrated with the queueing mechanism in htmx.  it has nothing to do with the hx-confirm attribute, which is a simple way to use a browser-based `confirm()` dialog.

the current sweet alert example that I'm familiar with uses a custom event:

https://htmx.org/examples/confirm/

and doesn't use the `htmx:confirm` event

abe-101 commented:
> the `htmx:confirm` event is always triggered on every request. this is by design: it provides a scripting mechanism to cancel or delay a request that is integrated with the queueing mechanism in htmx. it has nothing to do with the hx-confirm attribute, which is a simple way to use a browser-based `confirm()` dialog.
> 
> the current sweet alert example that I'm familiar with uses a custom event:
> 
> https://htmx.org/examples/confirm/
> 
> and doesn't use the `htmx:confirm` event

Can you clarify where on the page you are reffering to?

the example shown at https://htmx.org/examples/confirm/#vanilla-js-hx-confirm
does use the `htmx:confirm`, am I missing something?
